src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeReader.cs

```diff
@@ -188,74 +189,168 @@ public override void Complete(Exception? exception = null)
-                        await InnerStream.ReadAsync(Memory<byte>.Empty, cancellationToken).ConfigureAwait(false);
-                    }
+                        // This optimization only makes sense if we don't have anything buffered
+                        if (reader.UseZeroByteReads && reader._bufferedBytes == 0)
+                        {
+                            // Wait for data by doing 0 byte read before
+                            await reader.InnerStream.ReadAsync(Memory<byte>.Empty, tokenSource.Token).ConfigureAwait(false);
+                        }
 
-                    AllocateReadTail();
+                        reader.AllocateReadTail();
 
-                    Memory<byte> buffer = _readTail!.AvailableMemory.Slice(_readTail.End);
+                        Memory<byte> buffer = reader._readTail!.AvailableMemory.Slice(reader._readTail.End);
 
-                    int length = await InnerStream.ReadAsync(buffer, tokenSource.Token).ConfigureAwait(false);
+                        int length = await reader.InnerStream.ReadAsync(buffer, tokenSource.Token).ConfigureAwait(false);
 
-                    Debug.Assert(length + _readTail.End <= _readTail.AvailableMemory.Length);
+                        Debug.Assert(length + reader._readTail.End <= reader._readTail.AvailableMemory.Length);
 
-                    _readTail.End += length;
-                    _bufferedBytes += length;
+                        reader._readTail.End += length;
+                        reader._bufferedBytes += length;
 
-                    if (length == 0)
+                        if (length == 0)
+                        {
+                            reader._isStreamCompleted = true;
+                        }
+                    }
+                    catch (OperationCanceledException)
                     {
-                        _isStreamCompleted = true;
+                        reader.ClearCancellationToken();
+
+                        if (tokenSource.IsCancellationRequested && !cancellationToken.IsCancellationRequested)
+                        {
+                            // Catch cancellation and translate it into setting isCanceled = true
+                            isCanceled = true;
+                        }
+                        else
+                        {
+                            throw;
+                        }
+
                     }
+
+                    return new ReadResult(reader.GetCurrentReadOnlySequence(), isCanceled, reader._isStreamCompleted);
                 }
-                catch (OperationCanceledException)
+            }
+        }
+
+        protected override ValueTask<ReadResult> ReadAtLeastAsyncCore(int minimumSize, CancellationToken cancellationToken)
+        {
+            // TODO ReadyAsync needs to throw if there are overlapping reads.
+            ThrowIfCompleted();
+
+            cancellationToken.ThrowIfCancellationRequested();
```

Comment:
This should really be:
```C#
if (cancellationToken.IsCancellationRequested)
{
    return ValueTask.FromCanceled<ReadResult>(cancellationToken);
}
```
or to be able to target downlevel as well:
```C#
if (cancellationToken.IsCancellationRequested)
{
    return new ValueTask<ReadResult>(Task.FromCanceled<ReadResult>(cancellationToken));
}
```
Otherwise, this is going to propagate the OperationCanceledException synchronously out of the ReadAtLeastAsync call, when we generally prefer to propagate it as part of the returned task.

Reply:
@manandre would you like to take this one?

Reply:
Sure!
Fix proposal available in PR #53306